### PR TITLE
Synchronize access to peripherals array

### PR DESCRIPTION
**Pull request by filip-zielinski**
### Title
Synchronize access peripherals array

### Motivation
Synchronize access to an array of peripherals to prevent a race condition when using `BlueSwift` methods concurrently.
For example, BlueSwift could crash when `BluetoothConnection/disconnect(_:)` is called in a loop. Framework never declared such operation to be unsafe, hence the change, aiming to make it safe.

### Task Description
- implement `SynchronizedArray` data structure
- use `SynchronizedArray ` to internally store peripherals
- a block of read/write operations is called using `SynchronizedArray/read(_:)` and `SynchronizedArray/write(_:)`

### CODE REVIEW

File: Framework/Source Files/Utilities/SynchronizedArray.swift [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+/// To avoid race condition while performing multiple operations on ``SynchronizedArray`` use ``SynchronizedArray/read(_:)`` and ``SynchronizedArray/write(_:)`` methods.
+final class SynchronizedArray<Element> {
+
+    /// A concurrent queue which manages access to ``SynchronizedArray/storage``.
+    private let queue = DispatchQueue(label: "com.keto-mojo.sdk.framework.SynchronizedArray", attributes: .concurrent)
+    /// Internal storage of the ``SynchronizedArray``.
+    private var storage: [Element]
+
+    /// Initializes new ``SynchronizedArray`` with empty storage.
+    init() {
```
Comment (pkozielecki) [inline Framework/Source Files/Utilities/SynchronizedArray.swift:18]: will it be used in this module only?
if not, 🙏 consider making an object and init `public`

**Review by pkozielecki — APPROVED** [review]
Great work 👏 
Nice idea with that `synced array` 💪 
Added some suggestions, 🙏 👁️ ✋

File: Framework/Source Files/Utilities/SynchronizedArray.swift [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+//
+
+import Foundation
+
+/// A thread-safe `Array`, which synchronizes access to its elements.
+/// To avoid race condition while performing multiple operations on ``SynchronizedArray`` use ``SynchronizedArray/read(_:)`` and ``SynchronizedArray/write(_:)`` methods.
+final class SynchronizedArray<Element> {
+
+    /// A concurrent queue which manages access to ``SynchronizedArray/storage``.
+    private let queue = DispatchQueue(label: "com.keto-mojo.sdk.framework.SynchronizedArray", attributes: .concurrent)
```
Comment (esatkemal) [inline Framework/Source Files/Utilities/SynchronizedArray.swift:13]: I think we shouldn't expose Keto-Mojo name in this Repo `com.keto-mojo.sdk.framework.SynchronizedArray`. Maybe we can change the name it. WDYT?

**Review by esatkemal — COMMENTED** [review]
Great Job 🚀

File: Framework/Source Files/Utilities/SynchronizedArray.swift (hunk 2 above)
Comment (filip-zielinski) [reply Framework/Source Files/Utilities/SynchronizedArray.swift:13]: Yes, it's a leftover from my initial implementation in my app SDK. Good catch!

File: Framework/Source Files/Utilities/SynchronizedArray.swift [hunk 3]
```diff
@@ -0,0 +1,123 @@
+//
+//  Copyright © 2023 Netguru Sp. z o.o. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import Foundation
+
+/// A thread-safe `Array`, which synchronizes access to its elements.
+/// To avoid race condition while performing multiple operations on ``SynchronizedArray`` use ``SynchronizedArray/read(_:)`` and ``SynchronizedArray/write(_:)`` methods.
+final class SynchronizedArray<Element> {
```
[the remark this answers is not in the document]
Comment (filip-zielinski) [reply Framework/Source Files/Utilities/SynchronizedArray.swift:10]: Hmm, would it result in a performance gain?
From the perspective of the rest of the code base, It's not really needed to use only `Equatable` elements. We can always use `Equatable` elements on the call site.

File: Framework/Source Files/Utilities/SynchronizedArray.swift (hunk 1 above)
Comment (pkozielecki) [reply Framework/Source Files/Utilities/SynchronizedArray.swift:18]: 👍

File: Framework/Source Files/Utilities/SynchronizedArray.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (filip-zielinski) [reply Framework/Source Files/Utilities/SynchronizedArray.swift:10]: For the internal framework purpose, a `class` based implementation is enough. If `SynchronizedArray` were to become a separate framework, then value type semantic with copy-on-write would be desired, however it is not required here.

File: Bluetooth.xcodeproj/xcshareddata/xcschemes/BlueSwift-iOS.xcscheme [hunk 4]
```diff
@@ -54,7 +54,17 @@
       buildConfiguration = "Debug"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
-      shouldUseLaunchSchemeArgsEnv = "YES">
+      shouldUseLaunchSchemeArgsEnv = "YES"
+      codeCoverageEnabled = "YES">
```
Comment (filip-zielinski) [inline Bluetooth.xcodeproj/xcshareddata/xcschemes/BlueSwift-iOS.xcscheme:58]: Gather code coverage was enabled

File: Framework/Source Files/Utilities/SynchronizedArray.swift [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    /// A queue which manages access to ``SynchronizedArray/storage``. Should be a concurrent queue.
+    private let queue: DispatchQueueProtocol
```
Comment (filip-zielinski) [inline Framework/Source Files/Utilities/SynchronizedArray.swift:13]: Protocol to use in tests and to keep unit tests synchronous using `MockDispatchQueue`.

**Review by esatkemal — APPROVED** [review]
Great Work 🚀

**Review by pkozielecki — APPROVED** [review]
🚀

**Pull request merged**